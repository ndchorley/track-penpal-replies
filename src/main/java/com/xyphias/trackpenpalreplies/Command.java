package com.xyphias.trackpenpalreplies;

public sealed class Command permits ListPenpals, Quit {}
