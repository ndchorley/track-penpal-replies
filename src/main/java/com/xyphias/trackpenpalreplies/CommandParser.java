package com.xyphias.trackpenpalreplies;

public class CommandParser {
    public static Command parse(String command) {
        if (command.equals("L")) return new ListPenpals();
        
        return new Quit();
    }
}
