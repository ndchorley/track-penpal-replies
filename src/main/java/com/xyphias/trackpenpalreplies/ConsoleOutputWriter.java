package com.xyphias.trackpenpalreplies;

public class ConsoleOutputWriter implements OutputWriter {
    @Override
    public void writeLine(String line) {
        System.out.println(line);
    }

    @Override
    public void write(String string) {
        System.out.print(string);
    }
}
