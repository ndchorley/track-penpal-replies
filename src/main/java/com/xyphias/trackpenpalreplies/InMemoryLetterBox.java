package com.xyphias.trackpenpalreplies;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class InMemoryLetterBox implements LetterBox {
    public InMemoryLetterBox() {}

    @Override
    public boolean isEmpty() { return letters.isEmpty(); }

    @Override
    public List<Letter> contents() {
        return Collections.unmodifiableList(letters);
    }

    @Override
    public void add(Letter letter) {
        letters.add(letter);
    }

    private final ArrayList<Letter> letters = new ArrayList<>();
}
