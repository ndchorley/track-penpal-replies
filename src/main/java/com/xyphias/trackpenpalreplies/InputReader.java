package com.xyphias.trackpenpalreplies;

public interface InputReader {
    String readLine();
}
