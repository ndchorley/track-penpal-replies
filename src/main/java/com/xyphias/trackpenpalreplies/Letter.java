package com.xyphias.trackpenpalreplies;

import java.time.LocalDate;

public record Letter(Penpal from, LocalDate receivedOn) {}
