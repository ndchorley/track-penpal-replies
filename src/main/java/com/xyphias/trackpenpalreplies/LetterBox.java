package com.xyphias.trackpenpalreplies;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LetterBox {
    public LetterBox() {}

    public boolean isEmpty() { return letters.isEmpty(); }

    public List<Letter> contents() {
        return Collections.unmodifiableList(letters);
    }

    public void add(Letter letter) {
        letters.add(letter);
    }

    private final ArrayList<Letter> letters = new ArrayList<>();
}
