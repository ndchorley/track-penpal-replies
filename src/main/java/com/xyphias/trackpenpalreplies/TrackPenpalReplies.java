package com.xyphias.trackpenpalreplies;

import com.xyphias.trackpenpalreplies.commands.Command;
import com.xyphias.trackpenpalreplies.commands.CommandFactory;
import com.xyphias.trackpenpalreplies.commands.Quit;
import com.xyphias.trackpenpalreplies.infrastructure.io.InputReader;
import com.xyphias.trackpenpalreplies.infrastructure.io.OutputWriter;

public class TrackPenpalReplies {
    private final InputReader inputReader;
    private final OutputWriter outputWriter;
    private final LetterBox letterBox;
    private final CommandFactory commandFactory;

    public TrackPenpalReplies(
            LetterBox letterBox, InputReader inputReader, OutputWriter outputWriter, CommandFactory commandFactory
    ) {
        this.inputReader = inputReader;
        this.outputWriter = outputWriter;
        this.letterBox = letterBox;
        this.commandFactory = commandFactory;
    }

    public void run() {
        while (true) {
            Command command = readCommand();

            outputWriter.writeLine("");
            
            if (command instanceof Quit) break;
            
            command.execute();
        }
    }

    private Command readCommand() {
        outputWriter.write("enter command: ");

        String input = inputReader.readLine();

        Command command = commandFactory.createFrom(input);
        
        return command;
    }
}
