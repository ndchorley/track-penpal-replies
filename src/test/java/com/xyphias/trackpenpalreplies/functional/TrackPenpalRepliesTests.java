package com.xyphias.trackpenpalreplies.functional;

import com.xyphias.trackpenpalreplies.TrackPenpalReplies;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;

public class TrackPenpalRepliesTests {
    @Test
    public void it_displays_a_message_when_no_letters_need_a_reply() {
        List<String> commands = List.of("L", "Q");
        FakeInputReader inputReader = new FakeInputReader(commands);
        FakeOutputWriter outputWriter = new FakeOutputWriter();
        TrackPenpalReplies app = new TrackPenpalReplies(inputReader, outputWriter);

        app.run();

        Approvals.verify(outputWriter.written);
    }
}
