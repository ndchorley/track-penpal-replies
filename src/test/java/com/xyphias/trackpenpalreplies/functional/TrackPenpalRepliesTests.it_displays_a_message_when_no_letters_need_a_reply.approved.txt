enter command: 
No letters need a reply
enter command: 
